src/style/style_layer/fill_extrusion_style_layer.js

```diff
@@ -220,4 +253,96 @@ function projectExtrusion(geometry: Array<Array<Point>>, zBase: number, zTop: nu
+    const projectedBase = [];
+    const projectedTop = [];
+
+    for (const r of geometry) {
+        const ringBase = [];
+        const ringTop = [];
+        for (const p of r) {
+            const x = p.x + translation.x;
+            const y = p.y + translation.y;
+            const heightOffset = getTerrainHeightOffset(x, y, zBase, zTop, demSampler, centroid, exaggeration, lat);
+
+            const base = toPoint(vec3.transformMat4([], [x, y, heightOffset.base], m));
+            const top = toPoint(vec3.transformMat4([], [x, y, heightOffset.top], m));
+
+            ringBase.push(base);
+            ringTop.push(top);
+        }
+        projectedBase.push(ringBase);
+        projectedTop.push(ringTop);
+    }
+    return [projectedBase, projectedTop];
+}
+
+function toPoint(v: vec3): Point {
+    const p = new Point(v[0], v[1]);
+    p.z = v[2];
+    return p;
+}
+
+function getTerrainHeightOffset(x: number, y: number, zBase: number, zTop: number, demSampler: DEMSampler, centroid: vec2, exaggeration: number, lat: number): { base: number, top: number} {
+    const ele = exaggeration * demSampler.getElevationAt(x, y, true, true);
+    const flatRoof = centroid[0] !== 0;
+    const centroidElevation = flatRoof ? centroid[1] === 0 ? exaggeration * elevationFromUint16(centroid[0]) : exaggeration * flatElevation(demSampler, centroid, lat) : ele;
+    return {
+        base: ele + (zBase === 0) ? -1 : zBase, // Use -1 instead of -5 in shader to prevent picking underground
+        top: flatRoof ? Math.max(centroidElevation + zTop, ele + zBase + 2) : ele + zTop
+    };
+}
+
+function elevationFromUint16(n: number): number {
+    return n / 7.3;
+}
+
+function flatElevation(demSampler: DEMSampler, centroid: vec2, lat: number): number {
+    const pos = [Math.floor(centroid[0] / 8), Math.floor(centroid[1] / 8)];
+    const span = [10 * (centroid[0] - pos[0] * 8), 10 * (centroid[1] - pos[1] * 8)];
+
+    // Get height at centroid
+    const z = demSampler.getElevationAt(pos[0], pos[1], true, true);
+    const meterToDEM = demSampler.getMeterToDEM(lat);
+    const w = [
+        Math.floor(0.5 * (span[0] * meterToDEM - 1)),
+        Math.floor(0.5 * (span[1] * meterToDEM - 1))
+    ];
+    const posPx = demSampler.tileCoordToPixel(pos[0], pos[1]);
+
+    const offset = [2 * w[0] + 1, 2 * w[1] + 1];
+    const corners = fourSample(demSampler, [posPx.x - w[0], posPx.y - w[1]], offset);
+    const diff = [
+        Math.abs(corners[0] - corners[1]),
+        Math.abs(corners[2] - corners[3]),
+        Math.abs(corners[0] - corners[2]),
+        Math.abs(corners[1] - corners[3])
+    ];
+    const diffSum = [diff[0] + diff[1], diff[2] + diff[3]];
+    const slope = [
+        Math.min(0.25, meterToDEM * 0.5 * diffSum[0] / offset[0]),
+        Math.min(0.25, meterToDEM * 0.5 * diffSum[1] / offset[1])
+    ];
```

Comment:
Nit: a lot of temporary array allocations here that get immediately discarded. Should we inline some of these and keep more separate x/y variables?

Reply:
yea thats a good call it will make it more readable too